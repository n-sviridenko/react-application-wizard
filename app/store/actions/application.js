import { createAction } from 'redux-actions';

export const SHOW_REQUEST = 'app/application/SHOW_REQUEST';
export const SHOW_SUCCESS = 'app/application/SHOW_SUCCESS';
export const SHOW_FAIL = 'app/application/SHOW_FAIL';

export const showRequest = createAction(SHOW_REQUEST, null, (id) => ({ key: id }));
export const showSuccess = createAction(SHOW_SUCCESS, null, (item, id) => ({ key: id }));
export const showFail = createAction(SHOW_FAIL, null, (err, id) => ({ key: id }));

export const CREATE_REQUEST = 'app/application/CREATE_REQUEST';
export const CREATE_SUCCESS = 'app/application/CREATE_SUCCESS';
export const CREATE_FAIL = 'app/application/CREATE_FAIL';

export const createRequest = createAction(CREATE_REQUEST);
export const createSuccess = createAction(CREATE_SUCCESS);
export const createFail = createAction(CREATE_FAIL);
