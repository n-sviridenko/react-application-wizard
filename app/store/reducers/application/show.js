import * as schemas from 'store/schemas';
import itemReducerCreator, { createRootSelector } from 'store/reducers/creators/loaderCreator';
import listReducerCreator, { createItemSelector } from 'store/reducers/creators/listCreator';
import { SHOW_REQUEST, SHOW_SUCCESS, SHOW_FAIL } from 'store/actions/application';

export default listReducerCreator({
  itemReducerCreator,
  types: [
    SHOW_REQUEST,
    SHOW_SUCCESS,
    SHOW_FAIL,
  ],
});

const getRoot = (state) => state.getIn(['application', 'show']);
const makeGetRawById = (id) => createItemSelector(getRoot, id);
const makeGetById = (id) => createRootSelector(makeGetRawById(id), schemas.application);

export {
  getRoot,
  makeGetById,
};
