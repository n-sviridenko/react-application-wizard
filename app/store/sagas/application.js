import { call, put, takeLatest } from 'redux-saga/effects';

import * as schemas from 'store/schemas';
import * as actions from 'store/actions/application';
import { sendRequest } from './api';

export function* show({ payload: id, meta }) {
  try {
    const res = yield call(sendRequest, 'get', `applications/${id}`, null, schemas.application);

    yield put(actions.showSuccess(res, meta.key));
  } catch (err) {
    yield put(actions.showFail(err, meta.key));
  }
}

export function* showWatcher() {
  yield takeLatest(actions.SHOW_REQUEST, show);
}

export function* create({ payload }) {
  const data = {
    firstName: payload.get('firstName'),
    lastName: payload.get('lastName'),
    phone: payload.get('phone'),
    email: payload.get('email'),
    locationName: payload.get('location').name,
  };

  try {
    const res = yield call(sendRequest, 'post', 'applications', data, schemas.application);

    yield put(actions.createSuccess(res));
  } catch (err) {
    yield put(actions.createFail(err));
  }
}

export function* createWatcher() {
  yield takeLatest(actions.CREATE_REQUEST, create);
}

export default [
  showWatcher,
  createWatcher,
];
