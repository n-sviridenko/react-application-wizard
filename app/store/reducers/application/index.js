import { combineReducers } from 'redux-immutable';

import show from './show';

export default combineReducers({
  show,
});
