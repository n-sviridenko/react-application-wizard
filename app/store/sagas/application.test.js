/* eslint-disable redux-saga/yield-effects */
import { put, takeLatest } from 'redux-saga/effects';
import { Map } from 'immutable';

import * as actions from 'store/actions/application';
import { show, showWatcher, create, createWatcher } from './application';

describe('application saga', () => {
  describe('show', () => {
    let showTask;

    beforeEach(() => {
      showTask = show({ payload: 1, meta: { key: '1' } });

      const callDescriptor = showTask.next().value;
      expect(callDescriptor).toMatchSnapshot();
    });

    it('should dispatch the showSuccess if request was succeeded', () => {
      const putDescriptor = showTask.next(1).value;
      expect(putDescriptor).toEqual(put(actions.showSuccess(1, '1')));
    });

    it('should dispatch the showFail if request was failed', () => {
      const error = new Error();
      const putDescriptor = showTask.throw(error).value;
      expect(putDescriptor).toEqual(put(actions.showFail(error, '1')));
    });
  });

  describe('showWatcher', () => {
    const showWatcherTask = showWatcher();

    it('should start task to watch for SHOW_REQUEST action', () => {
      const takeLatestDescriptor = showWatcherTask.next().value;
      expect(takeLatestDescriptor).toEqual(takeLatest(actions.SHOW_REQUEST, show));
    });
  });

  describe('create', () => {
    let createTask;

    beforeEach(() => {
      createTask = create({
        payload: Map({
          firstName: 'Nikita',
          lastName: 'Sviridenko',
          email: 'mail.sviridenko@gmail.com',
          phone: null,
          location: {
            placeId: 'xxxx',
            name: 'Lyon, France',
          },
        }),
      });

      const callDescriptor = createTask.next().value;
      expect(callDescriptor).toMatchSnapshot();
    });

    it('should dispatch the createSuccess if request was succeeded', () => {
      const putDescriptor = createTask.next(1).value;
      expect(putDescriptor).toEqual(put(actions.createSuccess(1)));
    });

    it('should dispatch the createFail if request was failed', () => {
      const error = new Error();
      const putDescriptor = createTask.throw(error).value;
      expect(putDescriptor).toEqual(put(actions.createFail(error)));
    });
  });

  describe('createWatcher', () => {
    const createWatcherTask = createWatcher();

    it('should start task to watch for CREATE_REQUEST action', () => {
      const takeLatestDescriptor = createWatcherTask.next().value;
      expect(takeLatestDescriptor).toEqual(takeLatest(actions.CREATE_REQUEST, create));
    });
  });
});
