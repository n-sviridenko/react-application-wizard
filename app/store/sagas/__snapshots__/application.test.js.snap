exports[`application saga show should dispatch the showFail if request was failed 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "get",
      "applications/1",
      null,
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": "id",
        "_key": "application",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`application saga show should dispatch the showSuccess if request was succeeded 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "get",
      "applications/1",
      null,
      EntitySchema {
        "_getId": [Function],
        "_idAttribute": "id",
        "_key": "application",
        "_mergeStrategy": [Function],
        "_processStrategy": [Function],
        "schema": Object {},
      },
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
